syntax = "proto3";
package book;
option go_package = "bookpb";

message Book {
  string id = 1;
  string title = 2;
  string author = 3;
  repeated string genres = 4;
  double rating = 5;
}

message AddBookRequest { Book book = 1; }
message AddBookResponse { Book book = 1; }

message GetBookRequest { string id = 1; }
message GetBookResponse { Book book = 1; }

message ListBooksRequest {}
message ListBooksResponse { repeated Book books = 1; }

message SearchBooksRequest { string query = 1; }
message SearchBooksResponse { repeated Book books = 1; }

service BookCatalog {
  rpc AddBook(AddBookRequest) returns (AddBookResponse);
  rpc GetBook(GetBookRequest) returns (GetBookResponse);
  rpc ListBooks(ListBooksRequest) returns (ListBooksResponse);
  rpc SearchBooks(SearchBooksRequest) returns (SearchBooksResponse);
}
